shellcheck

https://github.com/koalaman/shellcheck

Output of pasting 7fa1172bd36937b55641be72c721f5586090a6a5 into https://www.shellcheck.net is below.

For details on issue reported (for example, SC2006), see https://github.com/koalaman/shellcheck/wiki/SC2006.

```
Line 16:
MYSQL=`which mysql`
      ^-- SC2006: Use $(..) instead of legacy `..`.
 
Line 17:
PSQL=`which psql`
     ^-- SC2006: Use $(..) instead of legacy `..`.
 
Line 18:
SQLITE3=`which sqlite3`
        ^-- SC2006: Use $(..) instead of legacy `..`.
 
Line 50:
  echo -e "$ME: $@" >&2
                ^-- SC2145: Argument mixes string and array. Use * or separate argument.
 
Line 68:
  printf "%s\n" "$VERSION"
            ^-- SC1117: Backslash is literal in "\n". Prefer explicit escaping: "\\n".
 
Line 170:
  if [ $? -ne 0 ]
       ^-- SC2181: Check exit code directly with e.g. 'if mycmd;', not indirectly with $?.
 
Line 224:
    "$MYSQL" $args <<< $*
             ^-- SC2086: Double quote to prevent globbing and word splitting.
                       ^-- SC2048: Use "$@" (with quotes) to prevent whitespace problems.
                       ^-- SC2086: Double quote to prevent globbing and word splitting.
 
Line 226:
    "$MYSQL" $args
             ^-- SC2086: Double quote to prevent globbing and word splitting.
 
Line 253:
cat >> $1 << EOF
       ^-- SC2086: Double quote to prevent globbing and word splitting.
 
Line 262:
cat >> $1 << EOF
       ^-- SC2086: Double quote to prevent globbing and word splitting.
 
Line 284:
    "$PSQL" $args -c "$*"
            ^-- SC2086: Double quote to prevent globbing and word splitting.
 
Line 286:
    "$PSQL" $args
            ^-- SC2086: Double quote to prevent globbing and word splitting.
 
Line 314:
cat >> $1 << EOF
       ^-- SC2086: Double quote to prevent globbing and word splitting.
 
Line 323:
cat >> $1 << EOF
       ^-- SC2086: Double quote to prevent globbing and word splitting.
 
Line 339:
    "$SQLITE3" $args <<< $*
               ^-- SC2086: Double quote to prevent globbing and word splitting.
                         ^-- SC2048: Use "$@" (with quotes) to prevent whitespace problems.
                         ^-- SC2086: Double quote to prevent globbing and word splitting.
 
Line 341:
    "$SQLITE3" $args
               ^-- SC2086: Double quote to prevent globbing and word splitting.
 
Line 369:
cat >> $1 << EOF
       ^-- SC2086: Double quote to prevent globbing and word splitting.
 
Line 378:
cat >> $1 << EOF
       ^-- SC2086: Double quote to prevent globbing and word splitting.
 
Line 424:
  local since="^-- $(sed_quote "$2")\$"
        ^-- SC2155: Declare and assign separately to avoid masking return values.
 
Line 425:
  local till="^-- $(sed_quote "$3")\$"
        ^-- SC2155: Declare and assign separately to avoid masking return values.
 
Line 453:
    | grep -q "\-$name\.sql" \
               ^-- SC1117: Backslash is literal in "\-". Prefer explicit escaping: "\\-".
                      ^-- SC1117: Backslash is literal in "\.". Prefer explicit escaping: "\\.".
 
Line 456:
  local fname="$MIGRATIONS/$(date +%s)-$(echo "$name" | tr \  _).sql"
        ^-- SC2155: Declare and assign separately to avoid masking return values.
 
Line 465:
${TYPE}_up_text $fname
^-- SC2086: Double quote to prevent globbing and word splitting.
                ^-- SC2086: Double quote to prevent globbing and word splitting.
 
Line 473:
${TYPE}_down_text $fname
^-- SC2086: Double quote to prevent globbing and word splitting.
                  ^-- SC2086: Double quote to prevent globbing and word splitting.
 
Line 484:
pending_migrations(){
^-- SC2120: pending_migrations references arguments, but none are ever passed.
 
Line 485:
  local last=$("${TYPE}_previous_versions" 1)
        ^-- SC2155: Declare and assign separately to avoid masking return values.
 
Line 504:
  while read version
        ^-- SC2162: read without -r will mangle backslashes.
 
Line 506:
    local migration=$(find_migrations -name "$version-*.sql")
          ^-- SC2155: Declare and assign separately to avoid masking return values.
 
Line 511:
  [[ $? -ne 0 ]] && exit 1
     ^-- SC2181: Check exit code directly with e.g. 'if mycmd;', not indirectly with $?.
 
Line 519:
  pending_migrations | while read fname
  ^-- SC2119: Use pending_migrations "$@" if function's $1 should mean script's $1.
                             ^-- SC2162: read without -r will mangle backslashes.
 
Line 577:
  while read fname
        ^-- SC2162: read without -r will mangle backslashes.
 
Line 579:
    local version=$(migration_version "$fname")
          ^-- SC2155: Declare and assign separately to avoid masking return values.
 
Line 580:
    local name=$(migration_name "$fname")
          ^-- SC2155: Declare and assign separately to avoid masking return values.
 
Line 593:
      echo -en "$ME: $message '${LMAGENTA}${name}${CLEAR}'\t(${LBLUE}$version${LBLUE}${CLEAR})... " >&2
                                                          ^-- SC1117: Backslash is literal in "\t". Prefer explicit escaping: "\\t".
 
Line 595:
      migration_section "$fname" "$since" "$till" || die "\n  ${LRED}migration in '${LYELLOW}$fname${LRED}' contains no section ${CYAN}$sname${CLEAR}"
                                                          ^-- SC1117: Backslash is literal in "\n". Prefer explicit escaping: "\\n".
 
Line 602:
  done < <("$src" $steps)
                  ^-- SC2086: Double quote to prevent globbing and word splitting.
 
Line 641:
shift $(($OPTIND - 1))
         ^-- SC2004: $/${} is unnecessary on arithmetic variables.
 
Line 649:
  source "$CONFIG"
  ^-- SC1090: Can't follow non-constant source. Use a directive to specify location.
 
Line 659:
[[ $ASK_PASSWORD -eq 1 ]] && read -s -p "Password: " _PASSWORD
                             ^-- SC2162: read without -r will mangle backslashes.
 
Line 694:
  RED="\e[31m"
       ^-- SC1117: Backslash is literal in "\e". Prefer explicit escaping: "\\e".

[ ... delete rest of SC1117 entries ...]
```

--- a/Makefile
+++ b/Makefile
@@ -7,6 +7,10 @@ install: ${PREFIX}/bin/shmig
 ${PREFIX}/bin/shmig: shmig
 	cp $? $@
 
-.PHONY: test
+.PHONY: test lint
+
 test:
 	(cd test ; make)
+
+lint:
+	shellcheck shmig

--- a/shmig
+++ b/shmig
@@ -1,4 +1,5 @@
 #!/usr/bin/env bash
+# shellcheck disable=SC2155
 
 # SHMIG. Database migration tool in BASH.
 # Copyright 2013 naquad <naquad@gmail.com>
@@ -13,9 +14,9 @@ CONFIG_EXPLICITLY_SET="0"
 ASK_PASSWORD="0"
 MIGRATIONS="./migrations"
 
-MYSQL=`which mysql`
-PSQL=`which psql`
-SQLITE3=`which sqlite3`
+MYSQL=$(command -v mysql)
+PSQL=$(command -v psql)
+SQLITE3=$(command -v sqlite3)
 
 UP_MARK="====  UP  ===="
 DOWN_MARK="==== DOWN ===="
@@ -28,26 +29,19 @@ COLOR="never"
 # we assume color output is not used
 
 RED=""
-GREEN=""
-YELLOW=""
-BLUE=""
-MAGENTA=""
 CYAN=""
-LGRAY=""
-DGRAY=""
 LRED=""
 LGREEN=""
 LYELLOW=""
 LBLUE=""
 LMAGENTA=""
 LCYAN=""
-LWHITE=""
 CLEAR=""
 BOLD=""
 
 # some helpers
 warn(){
-  echo -e "$ME: $@" >&2
+  echo -e "$ME: $*" >&2
 }
 
 die(){
@@ -166,8 +160,7 @@ __generic_checker(){
 
   TMPF=/tmp/$(((RANDOM<<15)|RANDOM)).$$.out
   rm -f $TMPF
-  "${TYPE}_cli" "$show" > $TMPF
-  if [ $? -ne 0 ]
+  if ! "${TYPE}_cli" "$show" > $TMPF
   then
     rm -f $TMPF
     return 1
@@ -212,18 +205,18 @@ __generic_status(){
 
 ##### MySQL
 mysql_cli(){
-  local args="-N -A -B $ARGS $DATABASE"
+  local args=(-N -A -B "$ARGS" "$DATABASE")
 
-  [[ -n $LOGIN    ]] && args="-u $LOGIN    $args"
-  [[ -n $HOST     ]] && args="-h $HOST     $args"
-  [[ -n $PORT     ]] && args="-P $PORT     $args"
+  [[ -n $LOGIN    ]] && args+=("-u" "$LOGIN")
+  [[ -n $HOST     ]] && args+=("-h" "$HOST")
+  [[ -n $PORT     ]] && args+=("-P" "$PORT")
   [[ -n $PASSWORD ]] && export MYSQL_PWD="$PASSWORD"
 
   if [[ $# -ne 0 ]]
   then
-    "$MYSQL" $args <<< $*
+    "$MYSQL" "${args[@]}" "$ARGS" "$DATABASE" <<< "$@"
   else
-    "$MYSQL" $args
+    "$MYSQL" "${args[@]}" "$ARGS" "$DATABASE"
   fi
 }
 
@@ -250,7 +243,7 @@ mysql_status(){
 }
 
 mysql_up_text() {
-cat >> $1 << EOF
+cat >> "$1" << EOF
 BEGIN;
 
 COMMIT;
@@ -259,7 +252,7 @@ EOF
 
 
 mysql_down_text() {
-cat >> $1 << EOF
+cat >> "$1" << EOF
 BEGIN;
 
 COMMIT;
@@ -272,18 +265,18 @@ EOF
 
 postgresql_cli(){
   export PGOPTIONS="-c client_min_messages=WARNING"
-  local args="-q -X -v VERBOSITY=terse -v ON_ERROR_STOP=1 -A -t -w -d $DATABASE $ARGS"
+  local args=(-q -X -v "VERBOSITY=terse" -v "ON_ERROR_STOP=1" -A -t -w -d "$DATABASE")
 
-  [[ -n $LOGIN    ]] && args="-U $LOGIN    $args"
-  [[ -n $HOST     ]] && args="-h $HOST     $args"
-  [[ -n $PORT     ]] && args="-p $PORT     $args"
+  [[ -n $LOGIN    ]] && args+=(-U "$LOGIN")
+  [[ -n $HOST     ]] && args+=(-h "$HOST")
+  [[ -n $PORT     ]] && args+=(-p "$PORT")
   [[ -n $PASSWORD ]] && export PGPASSWORD="$PASSWORD"
 
   if [[ $# -ne 0 ]]
   then
-    "$PSQL" $args -c "$*"
+    "$PSQL" "${args[@]}" "$ARGS" -c "$*"
   else
-    "$PSQL" $args
+    "$PSQL" "${args[@]}" "$ARGS"
   fi
 }
 
@@ -311,7 +304,7 @@ postgresql_status(){
 }
 
 postgresql_up_text() {
-cat >> $1 << EOF
+cat >> "$1" << EOF
 BEGIN;
 
 COMMIT;
@@ -320,7 +313,7 @@ EOF
 
 
 postgresql_down_text() {
-cat >> $1 << EOF
+cat >> "$1" << EOF
 BEGIN;
 
 COMMIT;
@@ -332,13 +325,13 @@ EOF
 ##### SQLite3
 
 sqlite3_cli(){
-  local args="-bail -batch $ARGS $DATABASE"
+  local args=(-bail -batch "$ARGS" "$DATABASE")
 
   if [[ $# -ne 0 ]]
   then
-    "$SQLITE3" $args <<< $*
+    "$SQLITE3" "${args[@]}" <<< "$@"
   else
-    "$SQLITE3" $args
+    "$SQLITE3" "${args[@]}"
   fi
 }
 
@@ -366,7 +359,7 @@ sqlite3_status(){
 }
 
 sqlite3_up_text() {
-cat >> $1 << EOF
+cat >> "$1" << EOF
 BEGIN;
 	PRAGMA foreign_keys = ON;
 
@@ -375,7 +368,7 @@ EOF
 }
 
 sqlite3_down_text() {
-cat >> $1 << EOF
+cat >> "$1" << EOF
 BEGIN;
 
 COMMIT;
@@ -462,7 +455,7 @@ create(){
 
 EOF
 
-${TYPE}_up_text $fname
+"${TYPE}_up_text" "$fname"
 
 cat >> "$fname" << EOF
 
@@ -470,7 +463,7 @@ cat >> "$fname" << EOF
 
 EOF
 
-${TYPE}_down_text $fname
+"${TYPE}_down_text" "$fname"
 
   local retval="$?"
 
@@ -481,6 +474,7 @@ ${TYPE}_down_text $fname
 
 # prints N pending migrations sorted by version in ascending order
 # if N is not given then all printed
+ # shellcheck disable=SC2120
 pending_migrations(){
   local last=$("${TYPE}_previous_versions" 1)
   last="${last:-0}"
@@ -501,13 +495,14 @@ pending_migrations(){
 previous_migrations(){
   local version=""
 
-  while read version
+  while read -r version
   do
     local migration=$(find_migrations -name "$version-*.sql")
     [[ -z $migration ]] && die "${LRED}migration version ${LYELLOW}$version ${LRED}can't be found${CLEAR}"
     echo "$migration"
   done < <("${TYPE}_previous_versions" "$1")
 
+  # shellcheck disable=SC2181
   [[ $? -ne 0 ]] && exit 1
 }
 
@@ -516,7 +511,8 @@ pending(){
   [[ $# -ne 0 ]] && die "${LRED}pending takes no arguments${CLEAR}"
 
   local fname=""
-  pending_migrations | while read fname
+  # shellcheck disable=SC2119
+  pending_migrations | while read -r fname
   do
     echo "${fname##*[/\\]}"
   done
@@ -574,7 +570,7 @@ migrate(){
   is_numeric "$stopver" || die "${LYELLOW}TILL${LRED} should be numeric${CLEAR}"
 
   local fname=""
-  while read fname
+  while read -r fname
   do
     local version=$(migration_version "$fname")
     local name=$(migration_name "$fname")
@@ -595,17 +591,17 @@ migrate(){
 
       migration_section "$fname" "$since" "$till" || die "\n  ${LRED}migration in '${LYELLOW}$fname${LRED}' contains no section ${CYAN}$sname${CLEAR}"
       "$version_change_sql" "$version"
-    ) | if "${TYPE}_cli" &> ${error_tmpfile}; then
+    ) | if "${TYPE}_cli" &> "${error_tmpfile}"; then
           echo -e "${BOLD}${LGREEN}done${CLEAR}" >&2
         else
           echo -e "${BOLD}${LRED}error${CLEAR}" >&2
-          cat ${error_tmpfile}
-          rm ${error_tmpfile}
+          cat "${error_tmpfile}"
+          rm "${error_tmpfile}"
           exit 1
         fi
 
     [[ ${PIPESTATUS[0]} -eq 0 && ${PIPESTATUS[1]} -eq 0 ]] || exit 1
-  done < <("$src" $steps)
+  done < <("$src" "$steps")
 }
 
 # a wrapper for migrate that applies migrations
@@ -644,7 +640,7 @@ do
     ?)   exit 1                    ;;
   esac
 done
-shift $(($OPTIND - 1))
+shift $((OPTIND - 1))
 
 # if config exists
 if [[ -e $CONFIG ]]
@@ -652,6 +648,7 @@ then
   # any error in configuration file should cause failure
   # to avoid potential database corruption
   trap 'die "Configuration failed"' ERR
+  # shellcheck source=/dev/null
   . "$CONFIG"
   # reset handler
   trap - ERR
@@ -668,13 +665,14 @@ if [ -e "$LOCAL_CONFIG" ]; then
   # any error in configuration file should cause failure
   # to avoid potential database corruption
   trap 'die "Local configuration failed"' ERR
+  # shellcheck source=/dev/null
   . "$LOCAL_CONFIG"
   # reset handler
   trap - ERR
 fi
 
 
-[[ $ASK_PASSWORD -eq 1 ]] && read -s -p "Password: " _PASSWORD
+[[ $ASK_PASSWORD -eq 1 ]] && read -r -s -p "Password: " _PASSWORD
 
 # options from command line have higher priority than configuration
 # file. if they were given then we overwrite values
@@ -710,20 +708,13 @@ esac
 if [[ $COLOR -eq 0 ]]
 then
   RED="\e[31m"
-  GREEN="\e[32m"
-  YELLOW="\e[33m"
-  BLUE="\e[34m"
-  MAGENTA="\e[35m"
   CYAN="\e[36m"
-  LGRAY="\e[37m"
-  DGRAY="\e[90m"
   LRED="\e[91m"
   LGREEN="\e[92m"
   LYELLOW="\e[93m"
   LBLUE="\e[94m"
   LMAGENTA="\e[95m"
   LCYAN="\e[96m"
-  LWHITE="\e[97m"
   BOLD="\e[1m"
   CLEAR="\e[0m"
 fi

